File: results/1.0/ZIP/zip-meta-llama-llama-3.3-70b-instruct-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

typedef struct {
    uint32_t signature;
    uint16_t version;
    uint16_t flags;
    uint16_t compression_method;
    uint16_t last_mod_file_time;
    uint16_t last_mod_file_date;
    uint32_t crc_32;
    uint32_t compressed_size;
    uint32_t uncompressed_size;
    uint16_t filename_length;
    uint16_t extra_field_length;
    uint8_t* filename;
    uint8_t* extra_field;
} LocalFileHeader;

typedef struct {
    uint32_t signature;
    uint32_t crc_32;
    uint32_t compressed_size;
    uint32_t uncompressed_size;
} DataDescriptor;

typedef struct {
    uint32_t signature;
    uint16_t version_made_by;
    uint16_t version_needed_to_extract;
    uint16_t flags;
    uint16_t compression_method;
    uint16_t last_mod_file_time;
    uint16_t last_mod_file_date;
    uint32_t crc_32;
    uint32_t compressed_size;
    uint32_t uncompressed_size;
    uint16_t filename_length;
    uint16_t extra_field_length;
    uint16_t comment_length;
    uint16_t disk_number_start;
    uint16_t internal_file_attributes;
    uint32_t external_file_attributes;
    uint32_t local_header_offset;
    uint8_t* filename;
    uint8_t* extra_field;
    uint8_t* comment;
} CentralDirectoryEntry;

typedef struct {
    uint32_t signature;
    uint16_t number_of_this_disk;
    uint16_t disk_where_central_directory_starts;
    uint16_t number_of_central_directory_entries_on_this_disk;
    uint16_t total_number_of_central_directory_entries;
    uint32_t size_of_central_directory;
    uint32_t offset_of_start_of_central_directory;
    uint16_t zipfile_comment_length;
    uint8_t* zipfile_comment;
} EndOfCentralDirectoryRecord;

typedef struct {
    LocalFileHeader* local_file_header;
    DataDescriptor* data_descriptor;
    uint8_t* file_data;
} File;

typedef struct {
    CentralDirectoryEntry** central_directory_entries;
    uint16_t number_of_central_directory_entries;
    EndOfCentralDirectoryRecord* end_of_central_directory_record;
} ZipFile;

#define HAMMER_PARSER_T struct hammer_parser_t
#define HAMMER_PARSER_TNX struct hammer_parser_tnx

HAMMER_PARSER_T* local_file_header_parser() {
    HAMMER_PARSER_T* signature = hammer_uint32_p(0x04034b50);
    HAMMER_PARSER_T* version = hammer_uint16_p();
    HAMMER_PARSER_T* flags = hammer_uint16_p();
    HAMMER_PARSER_T* compression_method = hammer_uint16_p();
    HAMMER_PARSER_T* last_mod_file_time = hammer_uint16_p();
    HAMMER_PARSER_T* last_mod_file_date = hammer_uint16_p();
    HAMMER_PARSER_T* crc_32 = hammer_uint32_p();
    HAMMER_PARSER_T* compressed_size = hammer_uint32_p();
    HAMMER_PARSER_T* uncompressed_size = hammer_uint32_p();
    HAMMER_PARSER_T* filename_length = hammer_uint16_p();
    HAMMER_PARSER_T* extra_field_length = hammer_uint16_p();

    HAMMER_PARSER_T* filename = hammer_string_p(filename_length);
    HAMMER_PARSER_T* extra_field = hammer_string_p(extra_field_length);

    LocalFileHeader* local_file_header = malloc(sizeof(LocalFileHeader));

    HAMMER_PARSER_T* parser = hammer_struct_p(&local_file_header->signature,
        &local_file_header->version,
        &local_file_header->flags,
        &local_file_header->compression_method,
        &local_file_header->last_mod_file_time,
        &local_file_header->last_mod_file_date,
        &local_file_header->crc_32,
        &local_file_header->compressed_size,
        &local_file_header->uncompressed_size,
        &local_file_header->filename_length,
        &local_file_header->extra_field_length,
        &local_file_header->filename,
        &local_file_header->extra_field);

    return parser;
}

HAMMER_PARSER_T* data_descriptor_parser() {
    HAMMER_PARSER_T* signature = hammer_uint32_p(0x08074b50);
    HAMMER_PARSER_T* crc_32 = hammer_uint32_p();
    HAMMER_PARSER_T* compressed_size = hammer_uint32_p();
    HAMMER_PARSER_T* uncompressed_size = hammer_uint32_p();

    DataDescriptor* data_descriptor = malloc(sizeof(DataDescriptor));

    HAMMER_PARSER_T* parser = hammer_struct_p(&data_descriptor->signature,
        &data_descriptor->crc_32,
        &data_descriptor->compressed_size,
        &data_descriptor->uncompressed_size);

    return parser;
}

HAMMER_PARSER_T* file_parser() {
    HAMMER_PARSER_T* local_file_header = local_file_header_parser();
    HAMMER_PARSER_T* data_descriptor = data_descriptor_parser();
    HAMMER_PARSER_T* file_data = hammer_string_p(100);

    File* file = malloc(sizeof(File));
    file->local_file_header = malloc(sizeof(LocalFileHeader));
    file->data_descriptor = malloc(sizeof(DataDescriptor));
    file->file_data = file_data;

    HAMMER_PARSER_T* parser = hammer_struct_p(&file->local_file_header->signature,
        &file->local_file_header->version,
        &file->local_file_header->flags,
        &file->local_file_header->compression_method,
        &file->local_file_header->last_mod_file_time,
        &file->local_file_header->last_mod_file_date,
        &file->local_file_header->crc_32,
        &file->local_file_header->compressed_size,
        &file->local_file_header->uncompressed_size,
        &file->local_file_header->filename_length,
        &file->local_file_header->extra_field_length,
        &file->local_file_header->filename,
        &file->local_file_header->extra_field,
        &file->data_descriptor->signature,
        &file->data_descriptor->crc_32,
        &file->data_descriptor->compressed_size,
        &file->data_descriptor->uncompressed_size,
        &file->file_data);

    return parser;
}

HAMMER_PARSER_T* central_directory_entry_parser() {
    HAMMER_PARSER_T* signature = hammer_uint32_p(0x02014b50);
    HAMMER_PARSER_T* version_made_by = hammer_uint16_p();
    HAMMER_PARSER_T* version_needed_to_extract = hammer_uint16_p();
    HAMMER_PARSER_T* flags = hammer_uint16_p();
    HAMMER_PARSER_T* compression_method = hammer_uint16_p();
    HAMMER_PARSER_T* last_mod_file_time = hammer_uint16_p();
    HAMMER_PARSER_T* last_mod_file_date = hammer_uint16_p();
    HAMMER_PARSER_T* crc_32 = hammer_uint32_p();
    HAMMER_PARSER_T* compressed_size = hammer_uint32_p();
    HAMMER_PARSER_T* uncompressed_size = hammer_uint32_p();
    HAMMER_PARSER_T* filename_length = hammer_uint16_p();
    HAMMER_PARSER_T* extra_field_length = hammer_uint16_p();
    HAMMER_PARSER_T* comment_length = hammer_uint16_p();
    HAMMER_PARSER_T* disk_number_start = hammer_uint16_p();
    HAMMER_PARSER_T* internal_file_attributes = hammer_uint16_p();
    HAMMER_PARSER_T* external_file_attributes = hammer_uint32_p();
    HAMMER_PARSER_T* local_header_offset = hammer_uint32_p();

    HAMMER_PARSER_T* filename = hammer_string_p(filename_length);
    HAMMER_PARSER_T* extra_field = hammer_string_p(extra_field_length);
    HAMMER_PARSER_T* comment = hammer_string_p(comment_length);

    CentralDirectoryEntry* central_directory_entry = malloc(sizeof(CentralDirectoryEntry));

    HAMMER_PARSER_T* parser = hammer_struct_p(&central_directory_entry->signature,
        &central_directory_entry->version_made_by,
        &central_directory_entry->version_needed_to_extract,
        &central_directory_entry->flags,
        &central_directory_entry->compression_method,
        &central_directory_entry->last_mod_file_time,
        &central_directory_entry->last_mod_file_date,
        &central_directory_entry->crc_32,
        &central_directory_entry->compressed_size,
        &central_directory_entry->uncompressed_size,
        &central_directory_entry->filename_length,
        &central_directory_entry->extra_field_length,
        &central_directory_entry->comment_length,
        &central_directory_entry->disk_number_start,
        &central_directory_entry->internal_file_attributes,
        &central_directory_entry->external_file_attributes,
        &central_directory_entry->local_header_offset,
        &central_directory_entry->filename,
        &central_directory_entry->extra_field,
        &central_directory_entry->comment);

    return parser;
}

HAMMER_PARSER_T* end_of_central_directory_record_parser() {
    HAMMER_PARSER_T* signature = hammer_uint32_p(0x06054b50);
    HAMMER_PARSER_T* number_of_this_disk = hammer_uint16_p();
    HAMMER_PARSER_T* disk_where_central_directory_starts = hammer_uint16_p();
    HAMMER_PARSER_T* number_of_central_directory_entries_on_this_disk = hammer_uint16_p();
    HAMMER_PARSER_T* total_number_of_central_directory_entries = hammer_uint16_p();
    HAMMER_PARSER_T* size_of_central_directory = hammer_uint32_p();
    HAMMER_PARSER_T* offset_of_start_of_central_directory = hammer_uint32_p();
    HAMMER_PARSER_T* zipfile_comment_length = hammer_uint16_p();

    HAMMER_PARSER_T* zipfile_comment = hammer_string_p(zipfile_comment_length);

    EndOfCentralDirectoryRecord* end_of_central_directory_record = malloc(sizeof(EndOfCentralDirectoryRecord));

    HAMMER_PARSER_T* parser = hammer_struct_p(&end_of_central_directory_record->signature,
        &end_of_central_directory_record->number_of_this_disk,
        &end_of_central_directory_record->disk_where_central_directory_starts,
        &end_of_central_directory_record->number_of_central_directory_entries_on_this_disk,
        &end_of_central_directory_record->total_number_of_central_directory_entries,
        &end_of_central_directory_record->size_of_central_directory,
        &end_of_central_directory_record->offset_of_start_of_central_directory,
        &end_of_central_directory_record->zipfile_comment_length,
        &end_of_central_directory_record->zipfile_comment);

    return parser;
}

int main(int argc, char** argv) {
    if (argc < 2) {
        printf("Usage: %s <zip_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (file == NULL) {
        printf("Error opening file: %s\n", argv[1]);
        return 1;
    }

    HAMMER_PARSER_T* parser = hammer_parser_init();

    ZipFile* zip_file = malloc(sizeof(ZipFile));

    HAMMER_PARSER_T* end_of_central_directory_record_parser_t = end_of_central_directory_record_parser();
    hammer_parse(parser, end_of_central_directory_record_parser_t, &zip_file->end_of_central_directory_record);

    CentralDirectoryEntry** central_directory_entries = malloc(sizeof(CentralDirectoryEntry*) * zip_file->end_of_central_directory_record->total_number_of_central_directory_entries);
    for (int i = 0; i < zip_file->end_of_central_directory_record->total_number_of_central_directory_entries; i++) {
        HAMMER_PARSER_T* central_directory_entry_parser_t = central_directory_entry_parser();
        hammer_parse(parser, central_directory_entry_parser_t, &central_directory_entries[i]);
    }
    zip_file->central_directory_entries = central_directory_entries;
    zip_file->number_of_central_directory_entries = zip_file->end_of_central_directory_record->total_number_of_central_directory_entries;

    fclose(file);

    printf("Number of central directory entries: %u\n", zip_file->number_of_central_directory_entries);

    for (int i = 0; i < zip_file->number_of_central_directory_entries; i++) {
        printf("Filename: %s\n", (char*)zip_file->central_directory_entries[i]->filename);
        printf("Compression method: %u\n", zip_file->central_directory_entries[i]->compression_method);
        printf("Compressed size: %u\n", zip_file->central_directory_entries[i]->compressed_size);
        printf("Uncompressed size: %u\n", zip_file->central_directory_entries[i]->uncompressed_size);
        printf("\n");
    }

    return 0;
}

File: results/1.0/ZIP/zip-meta-llama-llama-3.3-70b-instruct-turbo.rs
use nom::branch::alt;
use nom::bytes::complete::{take};
use nom::combinator::{map};
use nom::error::{ErrorKind, ParseError};
use nom::number::complete::{be_u16, be_u32};
use nom::sequence::{tuple};
use nom::IResult;
use std::env;
use std::fs::File;
use std::io::Read;
use std::path::Path;

#[derive(Debug, PartialEq)]
enum ZipError {
    InvalidSignature,
    InvalidCompressionMethod,
    InvalidVersion,
    UnexpectedEof,
}

impl std::error::Error for ZipError {}

impl std::fmt::Display for ZipError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            ZipError::InvalidSignature => write!(f, "Invalid signature"),
            ZipError::InvalidCompressionMethod => write!(f, "Invalid compression method"),
            ZipError::InvalidVersion => write!(f, "Invalid version"),
            ZipError::UnexpectedEof => write!(f, "Unexpected end of file"),
        }
    }
}

impl ParseError<&[u8]> for ZipError {
    fn from_error_kind(input: &[u8], kind: ErrorKind) -> Self {
        match kind {
            ErrorKind::AlphaNumeric => ZipError::InvalidSignature,
            ErrorKind::Digit => ZipError::InvalidCompressionMethod,
            ErrorKind::Eof => ZipError::UnexpectedEof,
            _ => ZipError::InvalidVersion,
        }
    }

    fn append(input: &[u8], kind: ErrorKind, error: ZipError) -> Self {
        error
    }
}

#[derive(Debug, PartialEq)]
struct LocalFileHeader {
    signature: u32,
    version_needed_to_extract: u16,
    general_purpose_bit_flag: u16,
    compression_method: u16,
    last_modified_time: u16,
    last_modified_date: u16,
    crc32: u32,
    compressed_size: u32,
    uncompressed_size: u32,
    file_name_length: u16,
    extra_field_length: u16,
}

#[derive(Debug, PartialEq)]
struct CentralDirectory {
    signature: u32,
    version_made_by: u16,
    version_needed_to_extract: u16,
    general_purpose_bit_flag: u16,
    compression_method: u16,
    last_modified_time: u16,
    last_modified_date: u16,
    crc32: u32,
    compressed_size: u32,
    uncompressed_size: u32,
    file_name_length: u16,
    extra_field_length: u16,
    file_comment_length: u16,
    disk_number_start: u16,
    internal_file_attributes: u16,
    external_file_attributes: u32,
    relative_offset_of_local_header: u32,
}

#[derive(Debug, PartialEq)]
struct DataDescriptor {
    signature: u32,
    crc32: u32,
    compressed_size: u32,
    uncompressed_size: u32,
}

#[derive(Debug, PartialEq)]
struct EndOfCentralDirectory {
    signature: u32,
    number_of_this_disk: u16,
    number_of_the_disk_where_the_central_directory_starts: u16,
    number_of_entries_in_the_central_directory_on_this_disk: u16,
    total_number_of_entries_in_the_central_directory: u16,
    size_of_the_central_directory: u32,
    offset_of_start_of_central_directory: u32,
    comment_length: u16,
}

fn parse_u32(input: &[u8]) -> IResult<&[u8], u32, ZipError> {
    let (input, bytes) = take(4usize)(input)?;
    let value = u32::from_be_bytes(bytes.try_into().unwrap());
    Ok((input, value))
}

fn parse_u16(input: &[u8]) -> IResult<&[u8], u16, ZipError> {
    let (input, bytes) = take(2usize)(input)?;
    let value = u16::from_be_bytes(bytes.try_into().unwrap());
    Ok((input, value))
}

fn parse_local_file_header(input: &[u8]) -> IResult<&[u8], LocalFileHeader, ZipError> {
    let (input, signature) = parse_u32(input)?;
    let (input, version_needed_to_extract) = parse_u16(input)?;
    let (input, general_purpose_bit_flag) = parse_u16(input)?;
    let (input, compression_method) = parse_u16(input)?;
    let (input, last_modified_time) = parse_u16(input)?;
    let (input, last_modified_date) = parse_u16(input)?;
    let (input, crc32) = parse_u32(input)?;
    let (input, compressed_size) = parse_u32(input)?;
    let (input, uncompressed_size) = parse_u32(input)?;
    let (input, file_name_length) = parse_u16(input)?;
    let (input, extra_field_length) = parse_u16(input)?;

    Ok((
        input,
        LocalFileHeader {
            signature,
            version_needed_to_extract,
            general_purpose_bit_flag,
            compression_method,
            last_modified_time,
            last_modified_date,
            crc32,
            compressed_size,
            uncompressed_size,
            file_name_length,
            extra_field_length,
        },
    ))
}

fn parse_central_directory(input: &[u8]) -> IResult<&[u8], CentralDirectory, ZipError> {
    let (input, signature) = parse_u32(input)?;
    let (input, version_made_by) = parse_u16(input)?;
    let (input, version_needed_to_extract) = parse_u16(input)?;
    let (input, general_purpose_bit_flag) = parse_u16(input)?;
    let (input, compression_method) = parse_u16(input)?;
    let (input, last_modified_time) = parse_u16(input)?;
    let (input, last_modified_date) = parse_u16(input)?;
    let (input, crc32) = parse_u32(input)?;
    let (input, compressed_size) = parse_u32(input)?;
    let (input, uncompressed_size) = parse_u32(input)?;
    let (input, file_name_length) = parse_u16(input)?;
    let (input, extra_field_length) = parse_u16(input)?;
    let (input, file_comment_length) = parse_u16(input)?;
    let (input, disk_number_start) = parse_u16(input)?;
    let (input, internal_file_attributes) = parse_u16(input)?;
    let (input, external_file_attributes) = parse_u32(input)?;
    let (input, relative_offset_of_local_header) = parse_u32(input)?;

    Ok((
        input,
        CentralDirectory {
            signature,
            version_made_by,
            version_needed_to_extract,
            general_purpose_bit_flag,
            compression_method,
            last_modified_time,
            last_modified_date,
            crc32,
            compressed_size,
            uncompressed_size,
            file_name_length,
            extra_field_length,
            file_comment_length,
            disk_number_start,
            internal_file_attributes,
            external_file_attributes,
            relative_offset_of_local_header,
        },
    ))
}

fn parse_data_descriptor(input: &[u8]) -> IResult<&[u8], DataDescriptor, ZipError> {
    let (input, signature) = parse_u32(input)?;
    let (input, crc32) = parse_u32(input)?;
    let (input, compressed_size) = parse_u32(input)?;
    let (input, uncompressed_size) = parse_u32(input)?;

    Ok((
        input,
        DataDescriptor {
            signature,
            crc32,
            compressed_size,
            uncompressed_size,
        },
    ))
}

fn parse_end_of_central_directory(input: &[u8]) -> IResult<&[u8], EndOfCentralDirectory, ZipError> {
    let (input, signature) = parse_u32(input)?;
    let (input, number_of_this_disk) = parse_u16(input)?;
    let (input, number_of_the_disk_where_the_central_directory_starts) = parse_u16(input)?;
    let (input, number_of_entries_in_the_central_directory_on_this_disk) = parse_u16(input)?;
    let (input, total_number_of_entries_in_the_central_directory) = parse_u16(input)?;
    let (input, size_of_the_central_directory) = parse_u32(input)?;
    let (input, offset_of_start_of_central_directory) = parse_u32(input)?;
    let (input, comment_length) = parse_u16(input)?;

    Ok((
        input,
        EndOfCentralDirectory {
            signature,
            number_of_this_disk,
            number_of_the_disk_where_the_central_directory_starts,
            number_of_entries_in_the_central_directory_on_this_disk,
            total_number_of_entries_in_the_central_directory,
            size_of_the_central_directory,
            offset_of_start_of_central_directory,
            comment_length,
        },
    ))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: {} <file>", args[0]);
        return;
    }

    let path = Path::new(&args[1]);
    let mut file = match File::open(path) {
        Ok(file) => file,
        Err(err) => {
            println!("Error opening file: {}", err);
            return;
        }
    };

    let mut data = Vec::new();
    match file.read_to_end(&mut data) {
        Ok(_) => (),
        Err(err) => {
            println!("Error reading file: {}", err);
            return;
        }
    }

    let result = parse_local_file_header(&data);
    match result {
        Ok((_, lfh)) => {
            println!("Local File Header:");
            println!("Signature: 0x{:x}", lfh.signature);
            println!("Version needed to extract: {}", lfh.version_needed_to_extract);
            println!("General purpose bit flag: 0x{:x}", lfh.general_purpose_bit_flag);
            println!("Compression method: {}", lfh.compression_method);
            println!("Last modified time: {}", lfh.last_modified_time);
            println!("Last modified date: {}", lfh.last_modified_date);
            println!("CRC-32: 0x{:x}", lfh.crc32);
            println!("Compressed size: {}", lfh.compressed_size);
            println!("Uncompressed size: {}", lfh.uncompressed_size);
            println!("File name length: {}", lfh.file_name_length);
            println!("Extra field length: {}", lfh.extra_field_length);

            let mut offset = lfh.file_name_length as usize + lfh.extra_field_length as usize;
            while offset < data.len() {
                let result = parse_central_directory(&data[offset..]);
                match result {
                    Ok((_, cd)) => {
                        println!("Central Directory:");
                        println!("Signature: 0x{:x}", cd.signature);
                        println!("Version made by: {}", cd.version_made_by);
                        println!("Version needed to extract: {}", cd.version_needed_to_extract);
                        println!("General purpose bit flag: 0x{:x}", cd.general_purpose_bit_flag);
                        println!("Compression method: {}", cd.compression_method);
                        println!("Last modified time: {}", cd.last_modified_time);
                        println!("Last modified date: {}", cd.last_modified_date);
                        println!("CRC-32: 0x{:x}", cd.crc32);
                        println!("Compressed size: {}", cd.compressed_size);
                        println!("Uncompressed size: {}", cd.uncompressed_size);
                        println!("File name length: {}", cd.file_name_length);
                        println!("Extra field length: {}", cd.extra_field_length);
                        println!("File comment length: {}", cd.file_comment_length);
                        println!("Disk number start: {}", cd.disk_number_start);
                        println!("Internal file attributes: 0x{:x}", cd.internal_file_attributes);
                        println!("External file attributes: 0x{:x}", cd.external_file_attributes);
                        println!("Relative offset of local header: {}", cd.relative_offset_of_local_header);

                        offset += 46;
                    }
                    Err(err) => {
                        println!("Error parsing Central Directory: {:?}", err);
                        break;
                    }
                }
            }

            let result = parse_end_of_central_directory(&data[data.len() - 22..]);
            match result {
                Ok((_, eocd)) => {
                    println!("End of Central Directory:");
                    println!("Signature: 0x{:x}", eocd.signature);
                    println!("Number of this disk: {}", eocd.number_of_this_disk);
                    println!("Number of the disk where the central directory starts: {}", eocd.number_of_the_disk_where_the_central_directory_starts);
                    println!("Number of entries in the central directory on this disk: {}", eocd.number_of_entries_in_the_central_directory_on_this_disk);
                    println!("Total number of entries in the central directory: {}", eocd.total_number_of_entries_in_the_central_directory);
                    println!("Size of the central directory: {}", eocd.size_of_the_central_directory);
                    println!("Offset of start of central directory: {}", eocd.offset_of_start_of_central_directory);
                    println!("Comment length: {}", eocd.comment_length);
                }
                Err(err) => {
                    println!("Error parsing End of Central Directory: {:?}", err);
                }
            }
        }
        Err(err) => {
            println!("Error parsing Local File Header: {:?}", err);
        }
    }
}